r2-core/src/test/java/test/r2/transport/http/client/TestAsyncPool.java

```diff
@@ -570,6 +580,389 @@ public void testWaitTimeStats() throws Exception
+    Assert.assertEquals(stats.getCheckedOut(), numberOfCheckouts);
+    Assert.assertEquals(stats.getIdleCount(), concurrency);
+    Assert.assertEquals(stats.getTotalCreated(), numberOfCheckouts+concurrency);
+    Assert.assertEquals(stats.getPoolSize(), numberOfCheckouts+concurrency);
+    Assert.assertEquals(stats.getTotalDestroyed(), 0);
+    Assert.assertEquals(stats.getTotalBadDestroyed(), 0);
+    Assert.assertEquals(stats.getTotalTimedOut(), 0);
+  }
+
+  /***
+   * This test case verifies that the correct number of waiters are timed out while waiting for object from the pool
+   *
+   *     Assumption: the channel pool max size is always bigger than the requested checkout size
+   *
+   *|----------A------------|---------------B---------------|---------------C--------------|-------------D--------------
+   *   A = In Phase A , N number of object checkout request to the pool when there are no tasks pending in the rate
+   *       limiter. A's Expected result = channel pool will create N number of new objects and check them out
+   *   B = In Phase B, O number of object checkout request again sent to the channel pool when the pool has already
+   *       checkout N number of objects, In this phase, the object creation inside the pool is blocked
+   *       and the rate limiter will Queue the creation requests once it reached its maximum concurrency configured.
+   *   C = Ih Phase C, P number of objects are returned to the pool which are created in Phase A, this will make
+   *       the number of waiter queue size to be O-P
+   *   D = In Phase D, A delay will be introduced to timeout the waiters and all the O-P waiters should be timed out.
+   *       After the delay the object creation will be unblocked and it should create aleast the concurrency number of
+   *       objects even though the waiters are timedout.
+   *
+   * @param numberOfCheckoutsInPhaseA the N number of checkout operations that will be performed in phase A
+   * @param numberOfCheckoutsInPhaseB the O number of checkout operations that will be performed in Phase B
+   * @param numbOfObjectsToBeReturnedInPhaseC the numeber of objects returned in Phase C
+   * @param poolSize size of the pool,
+   * @param concurrency concurrency of the rate limiter
+   */
+  @Test(dataProvider = "waiterTimeoutDataProvider")
+  public void testWaiterTimeout(int numberOfCheckoutsInPhaseA, int numberOfCheckoutsInPhaseB,
+      int numbOfObjectsToBeReturnedInPhaseC,
+      int poolSize, int concurrency, int waiterTimeout)
+  {
+    CreationBlockableSynchronousLifecycle blockableObjectCreator =
+        new CreationBlockableSynchronousLifecycle(numberOfCheckoutsInPhaseB, concurrency);
+    ScheduledExecutorService executor = Executors.newScheduledThreadPool(500);
+    RateLimiter rateLimiter = new ExponentialBackOffRateLimiter(0, 5000,
+        10, executor, concurrency);
+
+    ClockedExecutor clockedExecutor = new ClockedExecutor();
+    SettableClock settableClock = new SettableClock();
```

Comment:
ClockedExecutor should extend Clock, so no need to have two

Reply:
The settableClock is needed here to simulate the time.  or Am I missing something ?

Reply:
clockedExecutor should do the same, no? if you replace all settableClock with the clockedExecutor, should work the same

Reply:
Updated, had to make a small adjustment to the until time to consider the pool thread interval.